### PR TITLE
make reuse_mem optional

### PR DESCRIPTION
**Pull request by ashokei**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/ngraph/ngraph_graph.h [hunk 1]
```diff
@@ -202,6 +202,7 @@ class Graph : public Node {
         context_(context),
         enable_fprop_cache(enable_fprop_cache) {
     fprop_cache = std::make_shared<ngraph::FpropCache>();
+    is_reuse_mem = context.dev_type != mxnet::Context::kCPU;
```
Comment (mbrookhart) [inline src/ngraph/ngraph_graph.h]: Can't GPU utilize mem reuse?

Comment (diyessi) [reply src/ngraph/ngraph_graph.h]: Actually, as written, "Can nothing but CPU utilize memory reuse?"

Comment (ashokei) [reply src/ngraph/ngraph_graph.h]: we compile graph targeted at specific backend. Since this flag is graph level, we can choose if mem_reuse should apply or not. At this time, we only anticipate one backend that cant reuse mem : NNP. All other backends can reuse mxnet mem (eg: cpu, gpu etc). 

The above line will be changed to kNNP, it is currently set to kCPU just for CI testing.

Comment (ashokei) [reply src/ngraph/ngraph_graph.h]: @diyessi agree, we were expecting ngraph to handle this transparently. 
This PR is mxnet temporary fix, until ngraph handles NNP mem-reuse.

Comment (diyessi) [reply src/ngraph/ngraph_graph.h]: Please add a JIRA so it gets addressed. JIRA remembers things better than I do.

**Comment by mbrookhart** [comment]
TESTNOW
